metrics/tags.go

```diff
@@ -3,70 +3,143 @@ package metrics
+// these tags, if it exists, or create a new TagSet with them and return it.
+func (ts *TagSet) With(name, value string) *TagSet {
+	return (*TagSet)(((*atlas.Node)(ts)).AddLink(name, value))
 }
 
-// TagSetFromSampleTags creates a TagSet starting
-// from the set defined by SampleTags.
-// Adding a tag on the new set doesn't impact the SampleTags.
-func TagSetFromSampleTags(st *SampleTags) *TagSet {
-	tm := &TagSet{}
-	tm.tags = st.tags
-	return tm
+// Without returns another TagSet, either an already existing one or a newly
+// created one, with all of the tags from the existing TagSet except the one
+// with the given key.
+func (ts *TagSet) Without(name string) *TagSet {
+	return (*TagSet)(((*atlas.Node)(ts)).DeleteKey(name))
 }
 
-// AddTag adds a tag in the set.
-func (tg *TagSet) AddTag(k, v string) {
-	tg.tags = tg.tags.AddLink(k, v)
+// Get returns the value of the tag with the given name and true, if that tag
+// exists in the set, and an empty string and false otherwise.
+func (ts *TagSet) Get(name string) (string, bool) {
+	return ((*atlas.Node)(ts)).ValueByKey(name)
 }
 
-// Get returns the Tag value and true
-// if the provided key has been found.
-func (tg *TagSet) Get(k string) (string, bool) {
-	return tg.tags.ValueByKey(k)
+// Contains checks that each key=value tag pair in the provided TagSet exists in
+// the receiver tag set as well, i.e. that the given set is a sub-set of it.
+func (ts *TagSet) Contains(other *TagSet) bool {
+	return ((*atlas.Node)(ts)).Contains((*atlas.Node)(other))
 }
 
-// Len returns the number of the set keys.
-func (tg *TagSet) Len() int {
-	return tg.tags.Len()
+// IsEmpty checks if the tag set is empty, i.e. if it's the root atlas node.
+func (ts *TagSet) IsEmpty() bool {
+	return ((*atlas.Node)(ts)).IsRoot()
 }
 
-// Delete deletes the item related to the provided key.
-func (tg *TagSet) Delete(k string) {
-	tg.tags = tg.tags.DeleteKey(k)
+// Map returns a {key: value} string map with all of the tags in the set.
+func (ts *TagSet) Map() map[string]string {
+	return ((*atlas.Node)(ts)).Path()
 }
 
-// Map returns a map of string pairs with the items in the TagSet.
-func (tg *TagSet) Map() map[string]string {
-	return tg.tags.Path()
+// SortAndAddTags sorts the given tags by their keys and adds them to the
+// current tag set one by one, without branching out. This is generally
+// discouraged and sequential usage of the AddTag() method should be preferred
+// and used whenever possible.
+//
+// The only place this method should be used is if we already have a
+// map[string]string of tags, e.g. with the test-wide --tags from the root, with
+// scenario.tags, with custom per-reqeust tags from a user, etc. Then it's more
+// efficient to sort their keys before we add them. If we don't, go map
+// iteration happens in pseudo-random order and this will generate a lot of
+// useless dead-end atlas Nodes.
+func (ts *TagSet) SortAndAddTags(m map[string]string) *TagSet {
```

Comment:
Does anyone have any better ideas for this name? `WithSortedTags()` is my only one, but it's a bit misleading

Reply:
Maybe just `AddTags` - we don't have a not sort variant and arguably this is an implementation detail. 

We can still document it with explanation why. But arguably users should not care and in general the provided function should do the "write thing" even if their name doesn't spell it out.

Reply:
Then it should be `WithTags()`? :thinking:

Reply:
I guess to be similar ot the other names :+1: ... I do wonder though if someone won't decide to use that instead of `With` even when they don't need to :( 

How many places do we have where we use that? Can we maybe move it only there ... or at least move it out of the type - I think it will be better as a helper function and will make it more obvious that it should probably not be used. 

I am just a bit worried that somebody will have to add a tag and see `WithTags` and skip over `With`. Or have 2 tags and decide that instead of just calling `With` twice they will go with `WithTags` - effectively giving them self a speed bump.

So I kind of prefer if it's hard to find and definitely a thing you might read the documentation on.

Reply:
As you can see from https://github.com/grafana/k6/pull/2638/commits/eec1bc9afb712ba699034f6d27a5eacb56e21017, it is used in quite a few places, though mostly in the tests... Only in a few places in the normal k6 flow, though fairly crucial ones. I think it's very well documented and we know to check for its usage during code reviews, so it should be fine

Resolved by https://github.com/grafana/k6/pull/2638/commits/eec1bc9afb712ba699034f6d27a5eacb56e21017?

Reply:
>  we know to check for its usage during code reviews, so it should be fine


Unfortunately we are not the only people who will use it ... I am a lot less worried about us, but every extension developer who decides to emit a metric and add a tag will need to figure this out. And if we have two methods that do the same thing on the type I am worried they will use the "wrong" one.

Reply:
Also, now that you renamed it ... shouldn't it be `WithMap` or `WithTagsFromMap` ? Because `With` does add a tag as well ... that is the whole point.

Reply:
hmm :thinking: good point, :+1: for `WithTagsFromMap()`

Reply:
Resolved by the latest [force push](https://github.com/grafana/k6/compare/eec1bc9afb712ba699034f6d27a5eacb56e21017..bcc550324c84eb125e74870945ff33d9ce1314f8)